### ISSUE TITLE
Performance regression in lambdify

### ISSUE THREAD

**Issue by moorepants** [sympy/sympy#14671]
Offending PRs:

- https://github.com/sympy/sympy/pull/11312
- https://github.com/sympy/sympy/pull/12808
- https://github.com/sympy/sympy/pull/13822

Benchmarks:

- http://www.moorepants.info/misc/sympy-asv/#lambdify.TimeLambdifyEvaluate.time_lambdify_evaluate
- http://www.moorepants.info/misc/sympy-asv/#lambdify.TimeLambdifyCreation.time_lambdify_create

**Comment by asmeurer** [comment]
Another (probably unrelated) performance issue with lambdify https://github.com/sympy/sympy/issues/14654

**Comment by asmeurer** [comment]
#11312 was superseded by https://github.com/sympy/sympy/pull/12946

**Comment by moorepants** [comment]
I haven't evaluated the PRs that ASV points too, sometimes they are not quite right and manual checking is needed. Not sure why though.

**Comment by asmeurer** [comment]
The creation slowdown seems to be our old friend https://github.com/sympy/sympy/pull/12945 (notice that it's only slower in Python 2). The code from https://github.com/sympy/sympy/pull/13822 is just exasabating this.

**Comment by moorepants** [comment]
Good catch.

**Comment by asmeurer** [comment]
Here's a fix for the decorator https://github.com/sympy/sympy/pull/14683. It isn't much, but I don't know how to make it faster. 

Regarding the PythonCode PR, I don't understand how it affects the evaluation performance. I checked and the generated code for the benchmark is exactly the same. @bjodah any thoughts? 

@moorepants can you run the benchmarks on https://github.com/sympy/sympy/pull/14683?

**Comment by bjodah** [comment]
Was the code generated identical down to the letter, or were functions qualified with module names (which could slow things down)?

**Comment by asmeurer** [comment]
From debugging lambdify, here is `lstr` in master: `'lambda _Dummy_213,_Dummy_214,_Dummy_215,_Dummy_216,_Dummy_217,_Dummy_218,_Dummy_219,_Dummy_220,_Dummy_221,_Dummy_222,_Dummy_223,_Dummy_224,_Dummy_225,_Dummy_226,_Dummy_227,_Dummy_228,_Dummy_229,_Dummy_230,_Dummy_231,_Dummy_232,_Dummy_233,_Dummy_234,_Dummy_235,_Dummy_236,_Dummy_237,_Dummy_238,_Dummy_239,_Dummy_240: ([_Dummy_213 + _Dummy_214 - _Dummy_217 + 36.252574322669, _Dummy_213 - _Dummy_215 + _Dummy_216 + 21.3219379611249, _Dummy_216 + _Dummy_218 - _Dummy_219 + 9.9011158998744, 2*_Dummy_216 + _Dummy_218 - _Dummy_220 + 18.190422234653, 3*_Dummy_216 + _Dummy_218 - _Dummy_221 + 24.8679190043357, 4*_Dummy_216 + _Dummy_218 - _Dummy_222 + 29.9336062089226, 5*_Dummy_216 + _Dummy_218 - _Dummy_223 + 28.5520551531262, 2*_Dummy_213 - 2*_Dummy_217 - 2*_Dummy_218 + _Dummy_224 + 32.4401680272417, 3*_Dummy_214 + _Dummy_218 - _Dummy_225 + 34.9992934135095, 4*_Dummy_214 + _Dummy_218 - _Dummy_226 + 37.0716199972041, _Dummy_227 - _Dummy_228 + _Dummy_229 + 2*_Dummy_232 + 2*_Dummy_233 + 2*_Dummy_234 + 2*_Dummy_235 + 2*_Dummy_236 + 2*_Dummy_237 + 2*_Dummy_238 - _Dummy_239 - 2*_Dummy_240 - exp(_Dummy_213) + exp(_Dummy_214) - exp(_Dummy_215) - 2*exp(_Dummy_218) - 2*exp(_Dummy_219) - 2*exp(_Dummy_220) - 2*exp(_Dummy_221) - 2*exp(_Dummy_222) - 2*exp(_Dummy_223) - 2*exp(_Dummy_224) + exp(_Dummy_225) + 2*exp(_Dummy_226), -_Dummy_227 - _Dummy_228 - 4*_Dummy_229 - 3*_Dummy_230 - 2*_Dummy_231 - 3*_Dummy_233 - 6*_Dummy_234 - 9*_Dummy_235 - 12*_Dummy_236 - 15*_Dummy_237 - 2*_Dummy_238 - 3*_Dummy_239 - 4*_Dummy_240 + exp(_Dummy_213) + exp(_Dummy_214) + 4*exp(_Dummy_215) + 3*exp(_Dummy_216) + 2*exp(_Dummy_217) + 3*exp(_Dummy_219) + 6*exp(_Dummy_220) + 9*exp(_Dummy_221) + 12*exp(_Dummy_222) + 15*exp(_Dummy_223) + 2*exp(_Dummy_224) + 3*exp(_Dummy_225) + 4*exp(_Dummy_226), -_Dummy_229 - _Dummy_230 - _Dummy_233 - 2*_Dummy_234 - 3*_Dummy_235 - 4*_Dummy_236 - 5*_Dummy_237 + exp(_Dummy_215) + exp(_Dummy_216) + exp(_Dummy_219) + 2*exp(_Dummy_220) + 3*exp(_Dummy_221) + 4*exp(_Dummy_222) + 5*exp(_Dummy_223), -_Dummy_228 - _Dummy_231 - 2*_Dummy_238 - 3*_Dummy_239 - 4*_Dummy_240 + exp(_Dummy_214) + exp(_Dummy_217) + 2*exp(_Dummy_224) + 3*exp(_Dummy_225) + 4*exp(_Dummy_226), -_Dummy_232 - _Dummy_233 - _Dummy_234 - _Dummy_235 - _Dummy_236 - _Dummy_237 - 2*_Dummy_238 - _Dummy_239 - _Dummy_240 + exp(_Dummy_218) + exp(_Dummy_219) + exp(_Dummy_220) + exp(_Dummy_221) + exp(_Dummy_222) + exp(_Dummy_223) + 2*exp(_Dummy_224) + exp(_Dummy_225) + exp(_Dummy_226)])'`

and in 1.1.1: `'lambda _Dummy_19,_Dummy_20,_Dummy_21,_Dummy_22,_Dummy_23,_Dummy_24,_Dummy_25,_Dummy_26,_Dummy_27,_Dummy_28,_Dummy_29,_Dummy_30,_Dummy_31,_Dummy_32,_Dummy_33,_Dummy_34,_Dummy_35,_Dummy_36,_Dummy_37,_Dummy_38,_Dummy_39,_Dummy_40,_Dummy_41,_Dummy_42,_Dummy_43,_Dummy_44,_Dummy_45,_Dummy_46: ([36.252574322669 - _Dummy_23 + _Dummy_19 + _Dummy_20, 21.3219379611249 - _Dummy_21 + _Dummy_19 + _Dummy_22, 9.9011158998744 - _Dummy_25 + _Dummy_22 + _Dummy_24, 18.190422234653 - _Dummy_26 + 2*_Dummy_22 + _Dummy_24, 24.8679190043357 - _Dummy_27 + 3*_Dummy_22 + _Dummy_24, 29.9336062089226 - _Dummy_28 + 4*_Dummy_22 + _Dummy_24, 28.5520551531262 - _Dummy_29 + 5*_Dummy_22 + _Dummy_24, 32.4401680272417 - 2*_Dummy_23 - 2*_Dummy_24 + 2*_Dummy_19 + _Dummy_30, 34.9992934135095 - _Dummy_31 + 3*_Dummy_20 + _Dummy_24, 37.0716199972041 - _Dummy_32 + 4*_Dummy_20 + _Dummy_24, -_Dummy_34 - _Dummy_45 - exp(_Dummy_19) - exp(_Dummy_21) - 2*_Dummy_46 - 2*exp(_Dummy_24) - 2*exp(_Dummy_25) - 2*exp(_Dummy_26) - 2*exp(_Dummy_27) - 2*exp(_Dummy_28) - 2*exp(_Dummy_29) - 2*exp(_Dummy_30) + 2*_Dummy_38 + 2*_Dummy_39 + 2*_Dummy_40 + 2*_Dummy_41 + 2*_Dummy_42 + 2*_Dummy_43 + 2*_Dummy_44 + 2*exp(_Dummy_32) + _Dummy_33 + _Dummy_35 + exp(_Dummy_20) + exp(_Dummy_31), -_Dummy_33 - _Dummy_34 - 15*_Dummy_43 - 12*_Dummy_42 - 9*_Dummy_41 - 6*_Dummy_40 - 4*_Dummy_35 - 4*_Dummy_46 - 3*_Dummy_36 - 3*_Dummy_39 - 3*_Dummy_45 - 2*_Dummy_37 - 2*_Dummy_44 + 2*exp(_Dummy_23) + 2*exp(_Dummy_30) + 3*exp(_Dummy_22) + 3*exp(_Dummy_25) + 3*exp(_Dummy_31) + 4*exp(_Dummy_21) + 4*exp(_Dummy_32) + 6*exp(_Dummy_26) + 9*exp(_Dummy_27) + 12*exp(_Dummy_28) + 15*exp(_Dummy_29) + exp(_Dummy_19) + exp(_Dummy_20), -_Dummy_35 - _Dummy_36 - _Dummy_39 - 5*_Dummy_43 - 4*_Dummy_42 - 3*_Dummy_41 - 2*_Dummy_40 + 2*exp(_Dummy_26) + 3*exp(_Dummy_27) + 4*exp(_Dummy_28) + 5*exp(_Dummy_29) + exp(_Dummy_21) + exp(_Dummy_22) + exp(_Dummy_25), -_Dummy_34 - _Dummy_37 - 4*_Dummy_46 - 3*_Dummy_45 - 2*_Dummy_44 + 2*exp(_Dummy_30) + 3*exp(_Dummy_31) + 4*exp(_Dummy_32) + exp(_Dummy_20) + exp(_Dummy_23), -_Dummy_38 - _Dummy_39 - _Dummy_40 - _Dummy_41 - _Dummy_42 - _Dummy_43 - _Dummy_45 - _Dummy_46 - 2*_Dummy_44 + 2*exp(_Dummy_30) + exp(_Dummy_24) + exp(_Dummy_25) + exp(_Dummy_26) + exp(_Dummy_27) + exp(_Dummy_28) + exp(_Dummy_29) + exp(_Dummy_31) + exp(_Dummy_32)])'`

They look the same to me.

**Comment by caley** [comment]
As a crude fix, we could do something like this
```
from numpy import array

def f(x):
    try:
        return 1/x
    except ZeroDivisionError:
        builtin_numerics = (int, float, complex)
        if isinstance(x, builtin_numerics):
            x = array(x)

        return 1/x
```
i.e. only wrap the arguments if we run into a problem.  This would speed up this benchmark at least.  The overall benefit would depend on real world use: how often do zero divisions come up in (correct) code, or how often the arguments will already be numpy arrays, given that the user selected the numpy module.

**Comment by asmeurer** [comment]
The arguments will very often be numpy arrays, but not always. I can't say what the split would be but there are a good number of both. 

The problem in the [original issue](https://github.com/sympy/sympy/issues/11306) was a piecewise function that had `1/x` which normally wouldn't be evaluated, but because of the way the numpy expression works, all arguments are evaluated, even if the `np.select` doesn't choose it. 

Honestly I'm wondering if we should just remove it, and document that the input should be wrapped in `np.array` if ZeroDivisionError can occur. In other words, leave the wrapping to the user.

**Comment by caley** [comment]
I just noted in in #11306 that the numpy backend gets used by default, but the input wrapping is only enabled when numpy is explicitly requested, which can cause problems.  It would be reasonable to leave the wrapping to the user if they asked for the numpy module, but not if they're just expecting sensible default behaviour.

If the only reason for the argument wrapping is to make Piecewise work, then there is another option at the expense of quite complicated generated code.  The original example was
```
>>> p = sympy.Piecewise((1/x, y < -1), (x, y <= 1), (1/x, True))
>>> f = lambdify([x, y], p)
```
resulting in generated code
```
def _lambdifygenerated(x, y):
    return (select([less(y, -1),less_equal(y, 1),True], [1/x,x,1/x], default=numpy.nan))
```
Instead, we could generate something like
```
def _lambdifygenerated(x, y):
    def _piecewise_1():
        _condlist = []
        _choicelist = []

        _cond = y < -1
        if isscalar(_cond):
            if _cond:
                _condlist.append(_cond)
                _choicelist.append(1/x)

                return select(_condlist, _choicelist)
        else:
            _condlist.append(_cond)
            _choicelist.append(1/x)

        _cond = y <= 1
        if isscalar(_cond):
            if _cond:
                _condlist.append(_cond)
                _choicelist.append(x)

                return select(_condlist, _choicelist)
        else:
            _condlist.append(_cond)
            _choicelist.append(x)

        _cond = True
        if isscalar(_cond):
            if _cond:
                _condlist.append(_cond)
                _choicelist.append(1/x)

                return select(_condlist, _choicelist)
        else:
            _condlist.append(_cond)
            _choicelist.append(1/x)

        return select(_condlist, _choicelist)

    return _piecewise_1()
```
(isscalar is from numpy).  It is much more complicated, and would get more complicated if the expression contains more than one Piecewise, or nested Piecewise.  But still, it should be possible, and would also get rid of the warning messages, as in #14789.

Are there other reasons to wrap the input arguments, besides making Piecewise work?

**Comment by asmeurer** [comment]
I believe Piecewise is the only issue. The problem is that certain points are evaluated but never actually used. 

I think the ideal solution would be to construct an output array and use masks to only apply the piecewise functions at the conditions. This is possible now that we can generate full functions. This would be more efficient, as np.select evaluates every element of the input arrays at every piecewise condition, instead of only at the conditions that match. I'm not sure exactly how to construct an output array with the right shape.

**Comment by asmeurer** [comment]
Something like this

```py
>>> x, y = (np.array([0, 1]), np.array([0, -2]))
>>> b = np.broadcast(x, y)
>>> out = np.empty(b.shape)
>>> out[y < -1] = 1/x[y < -1]
>>> out[y < 1] = x[y < 1]
>>> out
array([ 0.,  1.])
```

(I think `np.broadcast` is the rig thing to use here)

**Comment by asmeurer** [comment]
Hmm, but we would have to apply the masks in reverse order, to ensure the first matching condition is applied. Meaning something like `Piecewise((log(-x), x < 0), (log(x), True))` from https://github.com/sympy/sympy/issues/14789 would still generate warnings.

**Comment by asmeurer** [comment]
I guess we can do something like

```py
mask1 = cond1
mask2 = cond2 & ~mask1
mask3 = cond3 & ~mask2
...
```

**Comment by caley** [comment]
If we're doing
```
>>> b = np.broadcast(x, y)
```
to figure out the output shape, we'd need to pull apart the expressions in each part of the Piecewise so that we know that we only need to broadcast on x, y.  Maybe that's doable in the example above, but for more complicated expressions?

**Comment by asmeurer** [comment]
Yeah, I was thinking about that too. What if there's an unused argument. It would affect the broadcasting. It is doable, though (just look at the `free_symbols`). 

Or maybe we can consider that `f = lambdify((x, y, z), ...)` as a numpy function should always broadcast x, y, and z, even if z is unused. This is sort of related to https://github.com/sympy/sympy/issues/5642.

**Comment by caley** [comment]
Yes, I agree about the wrapping.  I'll have a go at a pull request to remove it.  The code change will be simple enough (and I guess some tests will have to change), but it will be about getting the documentation right.

And yes, expecting all the input arguments to broadcast to a common shape seems like a reasonable way to define the behaviour.  If we did that, then for #5642, we could change the Numpy printer to print constants value val as
```
numpy.full(b.shape, val)
```
where b is the broadcast of the inputs (e.g. x, y, z,...).

On the other hand, if we call
```
lambdify((x, y, z), [x, y, z])
```
there's no need for x, y, z to have related shapes, and there might be real use cases where they don't.

So I'm undecided.

**Comment by asmeurer** [comment]
Ah right. If the lambdified expression is a list then each element should be broadcast independently. Let's open a new issue for the Piecewise thing. I think it will end up being more complicated to get right than initially anticipated.

**Comment by asmeurer** [comment]
I opened https://github.com/sympy/sympy/issues/15014 for the Piecewise thing. Can this issue be closed, or are there still unfixed performance regressions?

**Comment by oscarbenjamin** [comment]
@moorepants can you check the benchmarks on this? It looks as if it might be fixed.

**Comment by moorepants** [comment]
Yes, I just have to run the suite. It takes a few minutes to setup and then a day or two to run.

**Comment by oscarbenjamin** [comment]
@moorepants if you can find the time to run the benchmarks soon that would be great.

Otherwise I'll remove the milestone from this issue soon.

**Comment by oscarbenjamin** [comment]
I'm closing as fixed. @moorepants feel free to reopen if you see a problem in the benchmarks

**Comment by moorepants** [comment]
I don't have time to work on the benchmarks right now. I can get back up to speed over the winter break for school.

**Comment by oscarbenjamin** [comment]
I'll reopen this until it is confirmed as fixed
